specs/phase0/p2p-interface.md

```diff
@@ -1013,7 +1014,8 @@ Because Phase 0 does not have shards and thus does not have Shard Committees, th
 ```python
 def compute_subscribed_subnet(node_id: NodeID, epoch: Epoch, index: int) -> SubnetID:
     node_id_prefix = node_id >> (NODE_ID_BITS - ATTESTATION_SUBNET_PREFIX_BITS)
-    node_offset = node_id % EPOCHS_PER_SUBNET_SUBSCRIPTION
+    shuffling_prefix = node_id >> (NODE_ID_BITS - ATTESTATION_SUBNET_PREFIX_BITS - ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS)
+    node_offset = (shuffling_prefix +  (shuffling_prefix * (EPOCHS_PER_SUBNET_SUBSCRIPTION // 1 << ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS))) % EPOCHS_PER_SUBNET_SUBSCRIPTION
```

Comment:
This looks a bit confusing. 

I wanted to make the transition through the `EPOCHS_PER_SUBNET_SUBSCRIPTION` uniform. i.e All the nodes dont just transition in the first 8 epochs. 

So this spreads the jump uniformly throughout the duration. 

Its essentially `(x + x*32)%256`

The `node_id_prefix` mixes the start of transition so nodes of different subnets are not transitioning at the same time.

Reply:
> Its essentially (x + x*32)%256

Hm `(x + x*32)%256 == x * 33 % 256`. 
Shouldn't it be `x * 32 % 256` ?

I.e. we are effectively using 3 bits to stagger the shuffling by `0, 32, 64, ..., 224` epochs 

> `EPOCHS_PER_SUBNET_SUBSCRIPTION // 1 << ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS`

nit: Can be slightly reduced: 
`EPOCHS_PER_SUBNET_SUBSCRIPTION >> ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS`

Reply:
I am a bit confused by this logic, the shuffling prefix would be 9 bits and would range from (0,511). Do we want just the least significant 3 bits ( shuffling bits), to be ignored here ? As the node offset would be 0, till the shuffling prefix is 8. 

> I wanted to make the transition through the EPOCHS_PER_SUBNET_SUBSCRIPTION uniform. i.e All the nodes dont just transition in the first 8 epochs.

I am guessing this is the intention here of the logic  `(x + x*32)%256` ? 

@Nashatyrev 
If we use `x * 32 % 256` , we would have biased offsets since everything is a mutiple of 32. So you would miss all the values between the mutiples of 32 . 

@AgeManning  Can't we just do 
```
node_offset = min(node_offset,8) 
``` 
And leave this as it is ? 
```
    node_offset = shuffling_prefix % EPOCHS_PER_SUBNET_SUBSCRIPTION
```

Reply:
Thanks for the comments guys. 

Yeah this was confusing when I wrote it, which is why I'm hoping someone might help simplify. 

I've modified it slightly, hopefully its a bit clearer. I've modified the naming of some variables now and split the bits out separately. 

The `subnet_prefix` with current values is 6 bits and 3 bits for the shuffling. I was thinking of just using the 3 bits to transition the nodes, but figured we might as well use the extra 6 to help adjust the starting point of when the shuffling occurs. 

Lets assume `subnet_prefix` is x. Its essentially a number (0,511). 

If we use (x*32)%256, then all nodes on all subnets will transition on epoch 32,64,... etc. I think we want to avoid all nodes transitioning at the same time. 

The simplest way I could of think of doing this was to just translate the starting epoch by x. i.e (x + x*32)%256. I wrote it this way to show that there is a translation happening along with a scaling (i.e we scale by 32 to make it uniform across the duration). I've corrected this translation now.

Currently it reads:
```python
    subnet_prefix = node_id >> (NODE_ID_BITS - ATTESTATION_SUBNET_PREFIX_BITS)
    shuffling_bits = node_id >> (NODE_ID_BITS - ATTESTATION_SUBNET_PREFIX_BITS - ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS) % (1 << ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS)
    epoch_transition = (subnet_prefix +  (shuffling_bits * (EPOCHS_PER_SUBNET_SUBSCRIPTION >> ATTESTATION_SUBNET_SHUFFLING_PREFIX_BITS))) % EPOCHS_PER_SUBNET_SUBSCRIPTION
```


By adding `subnet_prefix` to `shuffling_bits*32`, the intention is that each subnet (which is defined by the prefix) will then have a different starting epoch. 

More explicitly with some examples. Lets say (in bits) subnet_prefix 000000 is on subnet 0 and subnet_prefix 000001 is on subnet 1. 

A node which is on subnet 0 (`subnet_prefix` of 000000) and has shuffling bits of 001 should transition on epoch 32. The next node on subnet 0 with shuffling bits 010 should then transition on epoch 64. 
A node which is on subnet 1 (`subnet_prefix` of 000001) and has shuffling bits of 001 should transition on epoch 33 and one with 010 on epoch 65. 

The idea of the translation (which I didn't do correctly the first time) was to get this staggered shift so all nodes don't transition at once.

Reply:
Thanks, it looks much more clear now 👍 

> By adding subnet_prefix to shuffling_bits*32, the intention is that each subnet (which is defined by the prefix) will then have a different starting epoch.

Oh, got it now! Totally makes sense to me 👍

Reply:
@nisdas - Sorry I forgot to reply to your comment. 

I think the current change I've made might answer the question. But I think we can't just use the `min()` function.  
Let me know if the current version still doesn't make sense or if you think there is some simplification we can make. I didn't fully follow how the min function might work here.

Reply:
@AgeManning It makes much more sense now with the clarification, I think you can ignore the `min()` comment, I was unclear on the original code which is why I suggested it. 

Only thing that comes to mind is that is there any reason to keep this to 8 transitions per subscription period ? Ex: Any reason we can't go to a larger number (64), this would allow for the network to even more gradually progress to different subnets or do we want to keep the shuffling bits small here (3 vs 6).

Reply:
Yeah, so the goal here is to try and be able to map a node-id prefix to any given subnet and any given epoch. So if we need to find nodes on a subnet, we can just search the prefix. 
The more bits we use for the shuffling the larger the prefix space we need. 
As it stands, we have 6 bits for each subnet, plus 3 bits for shuffling, giving a total of 9 bits. 
The mapping size is then 512 different prefixes map to individual subnets on a given epoch. If we add more bits in the size grows exponentially. We shouldn't have to calculate each mapping every epoch, but I dont think we want this to grow too large. I picked 3 bits because seemed like a decent middle ground.

Reply:
Fair enough, that makes sense then, thanks for the explanation :+1: